[#17] 통신을 위한 네트워크 관련 베이스 코드 작성

heechokim opened the pull request:
### 내용
- 서버와 통신하기 위한 네트워크 통신 베이스 코드 작업입니다.
- 레트로핏 라이브러리 추가합니다.
- 레트로핏 콜 어댑터 기능을 적용하여 서버팀과 합의한 응답 인터페이스에 따라 Success/Failure를 결정합니다.

### 작업 사항
- 폴더 구조는 아래처럼 생성했습니다.
~~~
network
ㄴ api
ㄴ callAdapter : 콜어댑터를 위해 필요한 클래스들을 포함하고 있습니다.
ㄴ client
~~~
- 주요 클래스
  - `Retrofit.kt` : ~base url는 임시로 github open api로 적용해두었습니다.~ base url은 임시로 포스트맨 가짜 서버 url로 적용해두었습니다. 추후 통신 작업 시 모여런 서버 base url로 변경 필요합니다.
  - `ApiResponse.kt` : 서버팀과 합의한 성공/실패 응답 인터페이스를 정의해둔 곳입니다.
  - `ApiResultCall` 클래스 : 레트로핏 Call 처리를 커스텀하게 처리하는 클래스입니다.
    - 아직 서버팀과 상세 에러 코드가 논의되지 않아, 200 외의 코드는 Failure로 처리해놓았습니다. 추후 변경이 필요합니다.

### 테스트
- 포스트맨 프로그램을 이용하여 가짜 서버를 생성 후, 테스트하였습니다.
- 성공 테스트
![image](https://user-images.githubusercontent.com/31889335/165101476-b6f2c1c0-fa71-4896-90e0-3de00ad30bb7.png)

- 실패 테스트
![image](https://user-images.githubusercontent.com/31889335/165101323-1c9c3b18-391d-4ef8-8830-f0a2ef19a7d5.png)
- 로그
![image](https://user-images.githubusercontent.com/31889335/165101550-28632ea8-5778-4014-bcf4-3c62ae0e73c8.png)


  
### 참고
- MainActivity에 usage를 작성해놓았습니다. 실제 통신 코드 작업에 들어갈 때에는 제거해야 합니다.

heechokim committed:
[#17] 통신을 위한 네트워크 관련 베이스 코드 작성
- 클래스, 파일 이름 변경

heechokim committed:
[#17] 통신을 위한 네트워크 관련 베이스 코드 작성
- 레트로핏 CallAdapter 적용

heechokim committed:
API 응답 인터페이스 적용

heechokim committed:
[#17] 통신을 위한 네트워크 관련 베이스 코드 작성
- 코루틴 버전 1.6.0 적용

heechokim commented on .idea/gradle.xml:
```diff
@@ -6,6 +6,7 @@
         <option name="testRunner" value="GRADLE" />
         <option name="distributionType" value="DEFAULT_WRAPPED" />
         <option name="externalProjectPath" value="$PROJECT_DIR$" />
+        <option name="gradleJvm" value="11" />
```
제거하기.

heechokim commented on .idea/misc.xml:
```diff
@@ -1,6 +1,23 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_11" default="true" project-jdk-name="Android Studio default JDK" project-jdk-type="JavaSDK">
+  <component name="DesignSurface">
+    <option name="filePathToZoomLevelMap">
+      <map>
+        <entry key="../../../../layout/custom_preview.xml" value="0.28078438087745944" />
+        <entry key="app/src/main/res/drawable-v24/ic_launcher_foreground.xml" value="0.155" />
+        <entry key="app/src/main/res/drawable/selector_rectangle_button.xml" value="0.15" />
+        <entry key="app/src/main/res/drawable/selector_rectangle_button_text.xml" value="0.15" />
+        <entry key="app/src/main/res/drawable/selector_round_button.xml" value="0.15" />
+        <entry key="app/src/main/res/drawable/selector_round_button_text.xml" value="0.15" />
+        <entry key="app/src/main/res/drawable/shape_default_edittext.xml" value="0.2" />
+        <entry key="app/src/main/res/drawable/shape_rectangle_button.xml" value="0.2" />
+        <entry key="app/src/main/res/drawable/shape_round_botton.xml" value="0.221" />
+        <entry key="app/src/main/res/layout/activity_main.xml" value="0.25470992500615963" />
+        <entry key="app/src/main/res/layout/view_badge_round_image.xml" value="0.3067708333333333" />
+      </map>
+    </option>
+  </component>
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_11" project-jdk-name="1.8" project-jdk-type="JavaSDK">
```
왜 올라갔을까 제거하기.

heechokim replied:
리베이스 잘못 했네요..

heechokim committed:
[#17] 통신을 위한 네트워크 관련 베이스 코드 작성
- 포스트맨 가짜 서버 적용 테스트

heechokim committed:
[#17] 통신을 위한 네트워크 관련 베이스 코드 작성
- 불필요하게 푸시된 파일 롤백

ethan-223 commented on app/build.gradle:
```diff
@@ -88,4 +89,11 @@ dependencies {
+
+    // Retrofit2
+    implementation 'com.squareup.retrofit2:retrofit:2.9.0'
+    implementation 'com.squareup.retrofit2:converter-gson:2.9.0'
```
중요한 건 아니지만, 두 의존성 버전은 계속 동일하게 갈 가능성이 높으니 묶어둘까요 ?
```suggestion
    // Retrofit2
    def retrofit_version = '2.9.0'
    implementation "com.squareup.retrofit2:$retrofit_version"
    implementation "com.squareup.retrofit2:converter-gson:$retrofit_version"
```

heechokim replied:
반영했습니다!

ethan-223 commented on app/src/main/java/com/moyerun/moyeorun_android/network/api/ApiService.kt:
```diff
@@ -0,0 +1,12 @@
+import com.moyerun.moyeorun_android.network.callAdapter.ApiResult
+import retrofit2.http.GET
+
+interface ApiService {
+    @GET("success")
+    suspend fun getSuccess(): ApiResult<Success<Any>>
+
+    @GET("failure")
+    suspend fun getFailure(): ApiResult<Success<Any>>
+}
```
서버 팀과 협의해서 Success, Error 데이터 클래스를 만들어두었는데요, Success 를 매번 이렇게 감싸야하는게 많이 아쉽습니다.
ResponseBodyConverter 를 사용해서 data 만 꺼내서 반환하는 건 어떨까요 ?

[[Android] Retrofit2 Converter를 커스텀 해보기](https://antdevhistory.tistory.com/52)

heechokim replied:
오 좋네요 저도 이 부분이 좀 아쉽다고 생각했는데 개선 방법이 떠오르지 않아 그냥 두었는데 참고 자료 보고 개선해보겠습니다!

heechokim replied:
이 부분 작업하다가 생긴 질문이 있는데요!
원하시는 모습이 Success로 감싸지는 부분이 빠지고 특정 데이터 클래스가 ApiResult<>로 감싸지는 모양일까요?
```suggestion
interface ApiService {
    @GET("success")
    suspend fun getSuccess(): ApiResult<User>

    @GET("failure")
    suspend fun getFailure(): ApiResult<Room>
}
```

ethan-223 replied:
넵 맞습니다 ~ 

`Success<User>` -> `User` 는 ResponseBodyConverter 가 수행할 거고,
`User` -> `ApiResult<User>` 는 CallAdapter 가 수행하겠네요

ethan-223 commented on app/src/main/java/com/moyerun/moyeorun_android/network/calladapter/ApiResult.kt:
```diff
@@ -0,0 +1,12 @@
+ */
+sealed class ApiResult<out T> {
+    data class Success<T>(val body: T) : ApiResult<T>()
+    data class Failure(val errorBody: Error) : ApiResult<Nothing>()
```
나중에 에러 내용에 따라 UI 상태를 변경해야한다면, Error 를 가지고 있는 것보다는 Throwable 을 가지고 있는 게 조금 더 유연하지 않을까요 ?

Error 를 가지고 있다면 `case` 필드로 비교를 해야하는데 ..
```kotlin
when (errorBody.case) {
  800 -> { ... }
  900 -> { ...}
  else -> { ...}
}
```
혹은 
```kotlin
enum ErrorCase(val case: String) {
    NOT_LOGIN("100"),
    ALEADY_STARTED_ROOM("600"),
    DUPLICATED_NICKNAME("700"),
    ....,
    UNKNOWN("999");

    companion object {
        fun findErrorCase(val errorBody: Error): ErrorCase {
             return values().find { case == errorBody.case }?: UNKNOWN
        }
}

when (ErrorCase.findErrorCase(errorbody.case)) {
    is NOT_LOGIN -> { ... }
    is ALEADY_STARTED_ROOM -> { ... }
    else -> { ... }
}
``` 
이런식으로 사용될 것 같습니다. 첫 번째 방법은 보시다시피 많이 번거롭고 에러 케이스 문자열 관리가 잘 안될 것 같구요, 두 번째 방법은 괜찮아보이지만 약간 아쉽습니다. 이유는 다음과 같습니다.

1. 매번 findErrorCase() 를 호출해야함
2. 발생해선 안되는 에러거나, 예의주시하고 있는 에러인 경우 Firebase 에 로깅하기 어렵다.

ApiResult.Failure 가 Throwable 을 필드로 갖고 있다면, CallAdapter 내에서 HttpException 이 발생한 경우, 미리 정의된 Exception 으로 감싸서 던질 수가 있습니다. 그럼 다음과 같이 바로 분기가 가능하죠.

```kotlin
when (error) {
    is AuthException -> { ... } // 로그인을 안한 상태, 토큰이 만료된 상태 등 대응 상태가 같은 여러 가지 에러들을 묶을 수 있음. 
    is RoomException -> { ... } // 세부적으로 나누고자 한다면 AlreadyStartedRoomException 으로 가도 됩니다.
    else -> { Lg.fe(error) } // 예상하지 못한 에러 (일어나선 안되는 에러)니까 Firebase에 로깅도 쉽게 됩니다.
}
```
이런 이유로, Throwable 을 사용해보는 건 어떨까요 ? 의견 자유롭게 주세요.

heechokim replied:
아하 코드로 보니까 저번에 말씀해주신 부분 까먹고 있었는데 기억도 나고 확실히 이해가 가네요! 가독성 측면이나 파베 로그 전송 면에서도 Throwable이 좋을 것 같아요~

heechokim replied:
에러를 Exception으로 Wrapping 하는 방식으로 Failure 부분 변경했습니다!

ethan-223 commented on app/src/main/java/com/moyerun/moyeorun_android/network/calladapter/ApiResultCall.kt:
```diff
@@ -0,0 +1,96 @@
+                        )
+                    } else {
+                        val message = "response body is null."
+                        val case = "001" // TODO : 바디가 null일 경우 합의된 케이스 번호로 변경 필요함. 001은 임시 번호임.
```
이런 경우는 Unknown 으로, "999" 같은 게 관례인듯해요 ㅋㅋ (뇌피셜) 아래에 에러 바디가 null 인 경우도 함께 Unknown 으로 다뤄도 될 것 같구요. 이 두 가지 케이스를 ui 로 다르게 보여줄 게 아니라면, 하나의 에러로 처리해도 좋아보입니다.

ethan-223 commented on app/src/main/java/com/moyerun/moyeorun_android/network/calladapter/ApiResultCall.kt:
```diff
@@ -0,0 +1,96 @@
+                            Response.success(ApiResult.Failure(Error(message, case)))
+                        )
+                    }
+                } else { // status code 200 아닌 다른 것들.
+                    val errorBody = response.errorBody()?.let { getErrorResponse(it) }
+                    if (errorBody != null) {
+                        callback.onResponse(
+                            this@ApiResultCall,
+                            Response.success(ApiResult.Failure(errorBody))
+                        )
+                    } else {
```
개인적으로 어떤 url 을 호출하다가 에러가 발생했는지도 같이 감싸졌으면 좋겠어요.
아마 `delegate.request().url().toString()` 로 어떤 Url 이었는지 알 수 있을거예요.

위에서 의견 드린대로 Throwable 로 만들어 던진다면, Throwable 밑에 NetworkException 을 만들고, url 필드를 만들 수 있지 않을까 생각이 드네요. 그 아래에 ApiException 을 만들고 errorBody 필드를 두거나 message, case 를 unwrapping 해서 가지고 있거나 .. 그 아래에 세부적인 에러들을 (AuthException, RoomException .. ) 두는 형태를 생각하고 있어요. 어떤가요 ? 

이렇게 한다면 errorBody 를 통해 어떤 Exception 을 던져야하는지 판별하는 유틸 클래스도 하나 필요하겠구요 !

heechokim replied:
url 까지 감싸져있는 것 좋은 것 같아요!

> 위에서 의견 드린대로 Throwable 로 만들어 던진다면, Throwable 밑에 NetworkException 을 만들고, url 필드를 만들 수 있지 않을까 생각이 드네요. 그 아래에 ApiException 을 만들고 errorBody 필드를 두거나 message, case 를 unwrapping 해서 가지고 있거나 .. 그 아래에 세부적인 에러들을 (AuthException, RoomException .. ) 두는 형태를 생각하고 있어요. 어떤가요 ?

그런데 요 내용은 말로만 들으니 조금 이해가 어려운데요! 간략한 코드로 설명해주실 수 있을까요? (귀찮게 해서 죄송,, 합니다 ㅋㅋ🙏🏻)

ethan-223 replied:
생각해봤는데 구조를 그렇게 깊게 가져갈 필요는 없을 것 같고, ApiException과 NetworkException 은 의미를 달리 가져가는 게 맞는 것 같습니다.

다음과 같은 구조 어떨까요 ?

```kotlin
// 서버에서 에러를 직접 내려준 경우. Http Status code 가 200 대가 아님 -> Error 로 파싱이 가능
class ApiException(val url: String, val error: Error): RuntimeException() {
    val case: String
        get() = error.case

    override val message: String?
        get() = error.message
}
```

```kotlin
// 서버에서 에러는 내려주는 것 외에 통신 과정에서 발생한 에러 -> Error 로 파싱이 불가능
class NetworkException(message: String, cause: Throwable) : RuntimeException(message, cause)
```
ApiException 의 경우, message 를 바로 보여주어야하는 경우가 있을 수 있기 때문에 url 과 message 를 따로 두고요, NetworkException 은 그렇지 않기 때문에 message 만 뒀습니다. message 에 url 을 그대로 써도 되니까요.

그래서 ApiResultCall 의 `!isSuccessful()` 부분을 고친다면

```kotlin
else { // status code 200 번대가 아님
    val errorBody = response.errorBody()

    if (errorBody != null) {
        val errorResponse = Gson().fromJson(errorBody.string(), Error::class.java)
        if (errorResponse != null) {
            callback.onResponse(
                this@ApiResultCall,
                Response.success(
                    ApiResult.Failure(
                        ApiErrorCase.getException(
                            requestUrl,
                            errorResponse,
                            HttpException(response)
                        )
                    )
                )
            )
            return
        }
    }
    callback.onResponse(
        this@ApiResultCall,
        Response.success(
            ApiResult.Failure(
                NetworkException(
                    requestUrl,
                    HttpException(response)
                )
            )
        )
    )
}

enum class ApiErrorCase(val case: String) {
    // 서버와 합의된 에러 case 들을 정의
    UNKNOWN("999");

    companion object {
        fun findError(case: String): ApiErrorCase {
            return values().find { it.case == case} ?: UNKNOWN
        }

        fun getException(url: String, error: Error, cause: Throwable? = null): Throwable {
            return when (findError(error.case)) {
                // 에러 case 마다 Exception 을 다대일 매핑.
                // 동일하게 대응할 것들을 묶을 수 있음 (ex. NOT_LOGIN, TOKEN_EXPIRED -> AuthException
                else -> ApiException(url, error)
            }
        }
    }
}
```

요런식으로 분기하는게 어떨까요?
errorBody 가 null 이 아니고 우리가 정의한 Error 타입으로 파싱이 된다면 ApiException 으로 다루고요, 그 외의 경우는 서버에 통신하긴 했지만 뭔가 잘못된 케이스일겁니다. 그럼 NetworkException 으로 다루고요.

마찬가지로 onFailure() 때에도 NetworkException 으로 퉁쳐도 좋을 것 같습니다.

```kotlin
override fun onFailure(call: Call<S>, t: Throwable) {
    callback.onResponse(
        this@ApiResultCall,
        Response.success(
            ApiResult.Failure(
                NetworkException(
                    call.request().url().toString(), t
                )
            )
        )
    )
}
```

결국엔 다 뜯어고치게 되는 것 같은데 .. 200번대에 들어왔지만 body() 가 null 인 경우에는 어떡할까요 ? 요것도 NetworkException 으로 해도 될 것 같아요. 
```kotlin
if (response.isSuccessful) {
    val body = response.body()
    if (body != null) {
        callback.onResponse(
            this@ApiResultCall,
            Response.success(ApiResult.Success(body))
        )
    } else {
        callback.onResponse(
            this@ApiResultCall,
            Response.success(
                ApiResult.Failure(
                    NetworkException(
                        "Response from $requestUrl was null " +
                                "but response body type was decleared as non-null"
                    )
                )
            )
        )
    }
}
```
여기선 그렇다할 cause 를 얻을 수 없으니, NetworkException의 message에 조금 더 자세한 정보를 남겨줍니다.
NullPointerException 을 만들어서 cause 로 둘 수 있으나, 그럼 메시지가 애매해져서요 .. 이렇게 해도 되지 않을까싶네요

heechokim replied:
와ㅓ... 갓 성규네요 어떻게 이렇게 구조를 따따딱 짤 수 있나요 심지어 이해도 잘 됨. ㄸㄷ 많이 배우네요 감사합니다. 참고해서 수정해볼게요!!

ethan-223 replied:
아 참고로 Error 로 파싱해서 매핑하는 걸 retrofit 의 responseBodyConverter 를 사용했더라구요. 이것도 CallAdapter 찾는 것과 유사하게 retrofit 에 등록된 responseBodyConverter 를 찾아서 매핑하는건데요, 지금은 Gson 을 등록해둬서 매핑이 되는겁니다.
그래서 다른 responseBodyConverter 도 추가될 수 있고 .. 어떤걸로 매핑을 해버릴지 모르는 위험성이 있으니 저는 그냥 `Gson().fromJson(errorBody.string(), Error::class.java)` 으로 변환했습니다. 코드도 더 짧고요 ~

heechokim replied:
requestUrl을 알 수 있게 기록하는 것과 Failure 시 Exception으로 던지도록 수정했습니다~

ethan-223 commented on app/src/main/java/com/moyerun/moyeorun_android/network/calladapter/ApiResultCall.kt:
```diff
@@ -0,0 +1,96 @@
+    override fun clone(): Call<ApiResult<S>> = ApiResultCall(delegate.clone())
+
+    override fun execute(): Response<ApiResult<S>> {
+        throw UnsupportedOperationException("NetworkResponseCall doesn't support execute")
```
파일 이름이 `ApiResultCall` 이라, 여기 메시지도 똑같이 가면 좋겠네요 ~

heechokim replied:
반영했습니다!

ethan-223 reviewed (commented):
요구 사항이 많아진 것 같은데.. 의견 주고받으면서 할 수 있는 만큼만 해봅시다. 
정 안되면 일단 합친 다음, 천천히 고도화해도 되니까요.

1. ResponseBodyConverter 사용
2. Failure 시 Throwable로 다루기
3. Exception 구조화
4. request url 기록

요정도로 피드백 드린 것 같은데, 숫자가 낮을 수록 우선순위가 높아보입니다. 의견 주세욥

heechokim commented:
피드백 감사합니다~ 성규님 네트워크 베이스 코드 로그인/회원가입 구현 시 필요한 거라면 언제까지는 머지 되어야 할까요? 날짜를 봐보고 먼저 합친 뒤 개선할지, 요 PR에서 수정할지 결정해보려 합니다~

ethan-223 commented:
auth 관련 다른 팀들이랑 싱크 맞춰볼게요. 근데 이부분 뿐만 아니라 프로필 사진 선택/편집 부분도 아직 전혀 개발이 안돼서, 어차피 데드라인 맞추기는 어려울듯하네요 😞

heechokim committed:
[#17] 통신을 위한 네트워크 관련 베이스 코드 작성
- 불필요하게 푸시된 파일 제거

heechokim committed:
[#17] 통신을 위한 네트워크 관련 베이스 코드 작성
- 리뷰 반영
- 레트로핏 버전 변수 정의
- Failure 시 Throwable로 다루도록 변경
- request url 기록
- Exception 구조화

ethan-223 commented on app/src/main/java/com/moyerun/moyeorun_android/common/exceptions/NetworkException.kt:
```diff
@@ -0,0 +1,7 @@
+
+import java.lang.RuntimeException
+
+class NetworkException(message: String, cause: Throwable) : RuntimeException(message, cause) {
+
+}
```
구현부가 비어있다면 { }는 생략해도 좋아보여요 ㅎㅎ

heechokim replied:
수정했습니다~!

ethan-223 commented on app/src/main/java/com/moyerun/moyeorun_android/network/calladapter/ApiResult.kt:
```diff
@@ -0,0 +1,10 @@
+package com.moyerun.moyeorun_android.network.calladapter
+
+/**
+ * Success : API 호출 성공 시, body를 Wrapping 합니다.
+ * Failure : API 호출 실패 시, errorBody를 Wrpping 합니다.
+ */
+sealed class ApiResult<out T> {
+    data class Success<T>(val body: T) : ApiResult<T>()
+    data class Failure(val exception: Throwable) : ApiResult<Nothing>()
+}
```
Failure 주석에 errorBody -> Throwable 로 수정이 필요해보입니다 ㅎ
그리고 중요한 건 아니지만 exception 과 throwable 은 상속관계라도 엄연히 throwable 이 더 큰 범위니까요, exception 보다는 throwable 네이밍이 좋아보입니다.
```suggestion
 * Success : API 호출 성공 시, body를 Wrapping 합니다.
 * Failure : API 호출 실패 시, throwable를 Wrapping 합니다.
 */
sealed class ApiResult<out T> {
    data class Success<T>(val body: T) : ApiResult<T>()
    data class Failure(val throwable: Throwable) : ApiResult<Nothing>()
}
```

heechokim replied:
좋네요오 수정했습니다~!

heechokim committed:
[#17] 통신을 위한 네트워크 관련 베이스 코드 작성 - 리뷰 반영
- 클래스 구현부가 비어있는 경우 중괄호를 제거합니다.
- 네이밍 변경합니다.

heechokim committed:
[#17] 통신을 위한 네트워크 관련 베이스 코드 작성 - 리뷰 반영2
- 통신 예제 코드 제거
- local.properties에 baseUrl 선언 및 적용

ethan-223 reviewed (approved):
고생하셨습니다
